[content-visibility] Add support for selection

rwlbuis opened the pull request:
#### e61b784a19ec8b6283afe2614730b9d4dfa79559
<pre>
[content-visibility] Add support for selection
<a href="https://bugs.webkit.org/show_bug.cgi?id=258194">https://bugs.webkit.org/show_bug.cgi?id=258194</a>

Reviewed by Tim Nguyen.

Make selection of skipped content a reason to make an element relevant to the user as specified here:
<a href="https://drafts.csswg.org/css-contain/#relevant-to-the-user">https://drafts.csswg.org/css-contain/#relevant-to-the-user</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-048-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-070-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-071-expected.txt:
* Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
(WebCore::ContentVisibilityDocumentState::updateRelevancyOfContentVisibilityElements):
* Source/WebCore/dom/Element.h:
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::setSelection):

Canonical link: <a href="https://commits.webkit.org/267474@main">https://commits.webkit.org/267474@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/46db0171ae31f57ae9ab615c5852efb6c7ac1737

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16661 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16984 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17431 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18443 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15622 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16851 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20214 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17125 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16859 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17246 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14423 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19228 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14490 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15107 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15479 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15273 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19569 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15872 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13474 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15062 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3996 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19429 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15700 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -94,6 +96,24 @@ IntersectionObserver* ContentVisibilityDocumentState::intersectionObserver(Docum
+        }
+    }
+    return false;
+}
```
Does something like this work? (walking down from the target rather than up from the intersecting node)

Maybe @rniwa / @whsieh / @megangardner have better suggestions though (we want to figure out if a given element has a selection inside)

```suggestion
static bool targetContainsSelection(const Element& target)
{
    auto selection = target.document().selection().selection();
    if (auto range = selection.range()) {
        for (Ref node : intersectingNodes(*range)) {
            if (target.containsIncludingShadowDOM(node))
                return true;
        }
    }
    return false;
}
```

nt1m replied:
@whsieh suggested using treeOrder() for more efficiency, here's something that probably works:

```
static bool targetContainsSelection(const Element& target)
{
    auto selectionRange = target.document().selection().selection().range();
    if (!selectionRange)
        return false;
    return contains<ComposedTree>(*selectionRange, makeBoundaryPointBeforeNodeContents(target)) || contains<ComposedTree>(*selectionRange, makeBoundaryPointAfterNodeContents(target));
}
```

rwlbuis replied:
Thanks guys! Will try this tomorrow (also looks like I need to rebase).

rwlbuis replied:
The snippet with contains fails on the selection tests.
I found this works:
return intersects<ComposedTree>(*selectionRange, target);
But I am not sure if that is more efficient than using intersectionNodes.

nt1m replied:
Let's use `return intersects<ComposedTree>(*selectionRange, target);`

It's less verbose and uses `treeOrder` as Wenson suggested.

rwlbuis replied:
Sure, done.

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2322,6 +2322,8 @@ void Document::updateLayoutIgnorePendingStylesheets(Document::RunPostLayoutTasks
             scheduleFullStyleRebuild();
     }
 
+    updateRelevancyOfContentVisibilityElements();
```
Can we leave out this change for #16311 ?

rwlbuis replied:
Removed.

nt1m commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -117,6 +120,16 @@ DidUpdateAnyContentRelevancy ContentVisibilityDocumentState::updateRelevancyOfCo
+                auto selectionRange = target.document().selection().selection().range();
+                if (!selectionRange)
+                    return false;
+                return intersects(*selectionRange, target);
```
```suggestion
                return intersects<ComposedTree>(*selectionRange, target);
```

Can we use the flat tree to match CSS here? That's not the default.

nt1m replied:
Also I realize this can just be `selectionRange && intersects<ComposedTree>(*selectionRange, target);`

rwlbuis replied:
Yeah, that is nicer, will fix before landing.

rwlbuis committed:
[content-visibility] Add support for selection
https://bugs.webkit.org/show_bug.cgi?id=258194

Reviewed by Tim Nguyen.

Make selection of skipped content a reason to make an element relevant to the user as specified here:
https://drafts.csswg.org/css-contain/#relevant-to-the-user

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-048-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-070-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-071-expected.txt:
* Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
(WebCore::ContentVisibilityDocumentState::updateRelevancyOfContentVisibilityElements):
* Source/WebCore/dom/Element.h:
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::setSelection):

Canonical link: https://commits.webkit.org/267474@main
